MetaMask/metamask-mobile
# feat: RNTar native modules (Android and iOS)

## **Description**

This PR adds a custom native module RNTar. This module takes a file system path to a `.tgz` (tar gzip) file. The module reads this file, decompresses it and places it in the target directory. It then returns the new location of the decompressed data. This PR is needed for the mobile snaps implementation and will be used in this following [PR](https://github.com/MetaMask/metamask-mobile/pull/7942).

- All of the javascript code is code fenced so non of it will appear in the main MetaMask app
- the native modules cannot be code fenced so they will be included in the main app. They are never called from the UI so it should not interfere.

Most of the native code (java and swift) was implemented and reviewed in previous prs. This change is bringing them into main.

- [Original Android PR](https://github.com/MetaMask/metamask-mobile/pull/6300)
- [Original iOS PR](https://github.com/MetaMask/metamask-mobile/pull/5926)

## **Related issues**

Progresses: https://github.com/MetaMask/accounts-planning/issues/143

## **Manual testing steps**
- clone this branch
- run `yarn setup`
- open `js.env`
- change `export METAMASK_BUILD_TYPE` to `export METAMASK_BUILD_TYPE="flask"`
- run `source .js.env`
- run `yarn watch:clean`
- run the app in ios or android
- navigate to settings/Snaps
- click the `Test RNTar` button
- it should succeed and log the output directory
- if you open the output directory (logged in the console) you should see the package folder with all of the code for the [filesnap](https://github.com/filecoin-project/filsnap)
- once you have confirmed that the module is properly decompressed, go back to `.js.env`
- change `METAMASK_BUILD_TYPE` back to `main`
- run `source .js.env`
- reload watchman (`yarn watch:clean`)
- navigate back to settings
- the `Snaps` section should no longer be present
- repeat these steps with the opposite operating system


### Running android unit tests
- open this branch in Android studio
- open the `RNTarTest.java`
- run the test suite
- it should pass

<img width="1728" alt="Screenshot 2023-11-29 at 11 45 27 PM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/982b899e-fc9c-46ed-9ffe-84af5e9c3f89">

## **Screenshots/Recordings**

### **After**

https://github.com/MetaMask/metamask-mobile/assets/22918444/2b95b1c1-54d7-4060-8f95-3d8d565b0203


https://github.com/MetaMask/metamask-mobile/assets/22918444/273be30e-c39d-4906-8828-04b8e93ca2af


## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
